pkg/cmd/build/build.go

```diff
@@ -0,0 +1,433 @@
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package build
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/images"
+	"github.com/containerd/containerd/images/archive"
+	"github.com/containerd/containerd/platforms"
+	dockerreference "github.com/containerd/containerd/reference/docker"
+	"github.com/containerd/nerdctl/pkg/api/types"
+	"github.com/containerd/nerdctl/pkg/buildkitutil"
+	"github.com/containerd/nerdctl/pkg/clientutil"
+	"github.com/containerd/nerdctl/pkg/platformutil"
+	"github.com/containerd/nerdctl/pkg/strutil"
+	"github.com/sirupsen/logrus"
+)
+
+func Build(ctx context.Context, options *types.BuildCommandOptions, stdin io.Reader, stdout, stderr io.Writer) error {
+	platform := options.Platform
+	buildctlBinary, buildctlArgs, needsLoading, metaFile, tags, cleanup, err := generateBuildctlArgs(ctx, stdin, options)
+	if err != nil {
+		return err
+	}
+	if cleanup != nil {
+		defer cleanup()
+	}
+
+	quiet := options.Quiet
+
+	logrus.Debugf("running %s %v", buildctlBinary, buildctlArgs)
+	buildctlCmd := exec.Command(buildctlBinary, buildctlArgs...)
+	buildctlCmd.Env = os.Environ()
+
+	var buildctlStdout io.Reader
+	if needsLoading {
+		buildctlStdout, err = buildctlCmd.StdoutPipe()
+		if err != nil {
+			return err
+		}
+	} else {
+		buildctlCmd.Stdout = stdout
+	}
+	if !quiet {
+		buildctlCmd.Stderr = stderr
+	}
+
+	if err := buildctlCmd.Start(); err != nil {
+		return err
+	}
+
+	if needsLoading {
+		platMC, err := platformutil.NewMatchComparer(false, platform)
+		if err != nil {
+			return err
+		}
+		if err = loadImage(ctx, buildctlStdout, options.GOptions.Namespace, options.GOptions.Address, options.GOptions.Snapshotter, stdout, platMC, quiet); err != nil {
+			return err
+		}
+	}
+
+	if err = buildctlCmd.Wait(); err != nil {
+		return err
+	}
+
+	iidFile := options.IidFile
+	if iidFile != "" {
+		id, err := getDigestFromMetaFile(metaFile)
+		if err != nil {
+			return err
+		}
+		if err := os.WriteFile(iidFile, []byte(id), 0600); err != nil {
+			return err
+		}
+	}
+
+	client, ctx, cancel, err := clientutil.NewClient(ctx, options.GOptions.Namespace, options.GOptions.Address)
+	if len(tags) > 1 {
+		logrus.Debug("Found more than 1 tag")
+		if err != nil {
+			return fmt.Errorf("unable to tag images: %s", err)
+		}
```

Comment:
The err is returned from `clientutil.NewClient`. Should move out of `if len(tags) > 1` and change the error message to something like "unable to create client".

Reply:
same for `defer cancel()` below. should be out of `if`

Reply:
https://github.com/containerd/nerdctl/pull/1780#discussion_r1059530536

Reply:
I mean we should move line 101 to within in the `if` block below (the same applies to line 190):

1. `client` only used in the `if`.
2. if `if` condition is `false`, `cancel` is not called.

If you see https://github.com/containerd/nerdctl/pull/1780#discussion_r1059530536, the two `clientutil.NewClient` lines are within the `if`

Reply:
my bad, Updated